src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj

```diff
@@ -42,6 +42,13 @@
             (assoc :pool-context pool-context)
             (assoc :borrow-timeout (:borrow-timeout config))
             (assoc :event-callbacks (atom []))))))
+  (stop
+   [this context]
+   (let [{:keys [pool-context]} (tk-services/service-context this)
+         flush-complete? (promise)]
+     (jruby-agents/send-flush-pool-for-shutdown! pool-context flush-complete?)
+     @flush-complete?)
```

Comment:
Maybe it would be good to log starting/completing messages at a DEBUG level around the dereference here so we can see from the log if this gets stuck for an exorbitant amount of time?